cmd/deployment/apm/create.go

```diff
@@ -18,59 +18,123 @@
+	Long:    apmCreateLong,
+	Example: apmCreateExample,
 	PreRunE: cobra.MaximumNArgs(0),
 	RunE: func(cmd *cobra.Command, args []string) error {
-		if err := cmdutil.FileOrStdin(cmd, "file"); err != nil {
-			return err
+		var track, _ = cmd.Flags().GetBool("track")
+		var generatePayload, _ = cmd.Flags().GetBool("generate-payload")
+		var zoneCount, _ = cmd.Flags().GetInt32("zones")
+		var size, _ = cmd.Flags().GetInt32("size")
+		var name, _ = cmd.Flags().GetString("name")
+		var refID, _ = cmd.Flags().GetString("ref-id")
+		var esRefID, _ = cmd.Flags().GetString("elasticsearch-ref-id")
+		var id, _ = cmd.Flags().GetString("id")
+		var version, _ = cmd.Flags().GetString("version")
+		var dt, _ = cmd.Flags().GetString("deployment-template")
+		var region string
+		if ecctl.Get().Config.Region == "" {
+			region = cmdutil.DefaultECERegion
 		}
 
-		f, err := input.NewFileOrReader(os.Stdin, cmd.Flag("file").Value.String())
-		if err != nil {
-			return err
+		var payload *models.ApmPayload
+		if err := cmdutil.FileOrStdin(cmd, "file"); err == nil {
+			err := cmdutil.DecodeDefinition(cmd, "file", &payload)
+			if err != nil && err != cmdutil.ErrNodefinitionLoaded {
+				return err
+			}
 		}
-		defer f.Close()
 
-		var definition models.CreateApmRequest
-		if err := json.NewDecoder(f).Decode(&definition); err != nil {
-			return err
+		if payload == nil {
+			p, err := depresource.NewApm(depresource.NewStateless{
+				DeploymentID:       id,
+				ElasticsearchRefID: esRefID,
+				API:                ecctl.Get().API,
+				RefID:              refID,
+				Version:            version,
+				Region:             region,
+				TemplateID:         dt,
+				Size:               size,
+				ZoneCount:          zoneCount,
+			})
+			if err != nil {
+				return err
+			}
+			payload = p
 		}
 
-		track, _ := cmd.Flags().GetBool("track")
-		res, err := apm.Create(apm.CreateParams{
-			API:              ecctl.Get().API,
-			ID:               cmd.Flag("id").Value.String(),
-			CreateApmRequest: definition,
-			TrackParams: util.TrackParams{
-				Track:  track,
-				Output: ecctl.Get().Config.OutputDevice,
+		// Returns the ApmPayload skipping the creation of the resources.
+		if generatePayload {
+			return ecctl.Get().Formatter.Format("", payload)
+		}
+
+		var updateParams = deployment.UpdateParams{
+			DeploymentID: id,
+			API:          ecctl.Get().API,
+			Request: &models.DeploymentUpdateRequest{
+				// Setting PruneOrphans to false since we don't want any side
+				// effects on the deployment when only a partial deployment
+				// definition is sent.
+				PruneOrphans: ec.Bool(false),
+				Name:         name,
+				Resources: &models.DeploymentUpdateResources{
+					Apm: []*models.ApmPayload{payload},
+				},
 			},
-		})
+		}
+
+		res, err := deployment.Update(updateParams)
 		if err != nil {
 			return err
 		}
-		return ecctl.Get().Formatter.Format(filepath.Join("apm", "create"), res)
+
+		if err := ecctl.Get().Formatter.Format("", res); err != nil {
```

Comment:
I have the same concern here about duplicating the tracking code in several places